### DISCUSSION TITLE
Can't trigger callback on Enter keypress, on HoldBrowser

### DISCUSSION THREAD

**Discussion by 64kramsystem** [fltk-rs/fltk-rs#1188] [Q&A] [answered]
Hello!

I'm not sure if this is a bug or not, so I'm posting here first.

I'm trying to trigger a callback on the HoldBrowser widget, but without success. Based on the documentation [here](https://www.fltk.org/doc-1.3/classFl__Hold__Browser.html) and [here](https://www.fltk.org/doc-1.3/classFl__Widget.html#ad4ef380ffb50f4d00fe93f7b6677e9e1), it seems that it should be possible, by setting the callbacks trigger to `FL_WHEN_ENTER_KEY|FL_WHEN_NOT_CHANGED`.

When I try this in `fltk-rs` though, it doesn't have any effect:

```rs
use fltk::{enums::*, prelude::*, *};

fn main() -> Result<(), FltkError> {
    let app = app::App::default();
    let mut win = window::Window::default().with_size(900, 300);
    let mut b = browser::HoldBrowser::new(10, 10, 900 - 20, 300 - 20, "");
    b.add("Item 1");
    b.add("Item 2");
    b.set_trigger(CallbackTrigger::EnterKey.union(CallbackTrigger::NotChanged));
    b.set_callback(|_browser| println!("callback invoked!"));
    win.end();
    win.show();
    app.run()
}
```

when Enter is tapped on an entry, nothing is printed.

Is this a limitation of FLTK, an issue with fltk-rs, or am I doing something wrong?

I've tried different `CallbackTrigger` combinations, but none worked. I'm also aware that one can use `handle()` to listen to events, but the callback solution, if working, is simpler.

**Comment by MoAlyousef** [comment, marked as answer]
Hello!

I think it's a limitation of FLTK's HoldBrowser, since the combination does work for input type widgets, and the equivalent C++ code acts  the same:
```c++
#include <FL/Fl.H>
#include <FL/Enumerations.H>
#include <FL/Fl_Hold_Browser.H>
#include <FL/Fl_Window.H>

int main() {
    auto win = new Fl_Window(400, 300);
    auto b = new Fl_Hold_Browser(0, 0, 400, 300);
    b->add("item 1");
    b->add("item 2");
    b->when(FL_WHEN_ENTER_KEY|FL_WHEN_NOT_CHANGED);
    b->callback([](Fl_Widget *, void *) { puts("works"); }, 0);
    win->end();
    win->show();
    return Fl::run();
}
```

The handle solution can also be simple:
```rust
    b.set_callback(|_browser| println!("callback invoked!"));
    b.handle(|b, ev| if ev == Event::KeyDown && app::event_key() == Key::Enter {
        b.do_callback();
        true
    } else {
        false
    });
```

**Reply by 64kramsystem** [reply]
Thanks! Yes, I'll go the `handle()` way :smile:
